lib/autoproj/daemon/package_repository.rb

```diff
@@ -55,7 +55,7 @@ def branch
                 explicit_branch = vcs[:remote_branch] || vcs[:branch]
                 return explicit_branch if explicit_branch
 
-                autobuild.importer.resolve_remote_head(
+                @branch ||= autobuild.importer.resolve_remote_head(
```

Comment:
This assumes the default branch won't change. Since the daemons are long-living, I think it might be a problem (I recently had that case: created a new repo, did not realize that the default branch was not master/main and had to change it)

Reply:
What do you suggest? The only way I see around the scenario you are pointing out is to _always_ fetch it from the remote. Even `Autobuild::Git` is assuming the default branch doesn't change (i.e if you change it you have to remove the package from the workspace and update).

Reply:
> I am assuming that the PackageRepository object is long-lived

Packages are loaded from the installation manifest once, when the daemon starts.

> Autobuild::Git in practice is re-created each time autoproj executes

Yeah but the remote HEAD ref is stored by git. [It will only fetch from the remote once when you clone the repository.](https://github.com/rock-core/autobuild/blob/master/lib/autobuild/import/git.rb#L348)

Reply:
Well, then, let's fix that too.

But let's not cache the same thing a second time when we know it's going to cause trouble.

Reply:
That makes sense but dynamic remote HEAD is not the feature I'm trying to implement right now. Even without this patch, changing the main branch is not supported. My goal right now is to avoid shelling out to git too frequently because:

```
  WARN: Exception raised by code in GitPoller#poll
  WARN: Waiting 10s and restarting the daemon
  WARN: Too many open files
  WARN:   /home/autoproj/.local/share/autoproj/gems/ruby/2.7.0/gems/autobuild-1.23.1/lib/autobuild/subcommand.rb:311:in `pipe'
  WARN:   /home/autoproj/.local/share/autoproj/gems/ruby/2.7.0/gems/autobuild-1.23.1/lib/autobuild/subcommand.rb:311:in `block in run'
  WARN:   /home/autoproj/.local/share/autoproj/gems/ruby/2.7.0/gems/autobuild-1.23.1/lib/autobuild/subcommand.rb:290:in `open'
  WARN:   /home/autoproj/.local/share/autoproj/gems/ruby/2.7.0/gems/autobuild-1.23.1/lib/autobuild/subcommand.rb:290:in `run'
  WARN:   /home/autoproj/.local/share/autoproj/gems/ruby/2.7.0/gems/autobuild-1.23.1/lib/autobuild/package.rb:586:in `run'
  WARN:   /home/autoproj/.local/share/autoproj/gems/ruby/2.7.0/gems/autobuild-1.23.1/lib/autobuild/import/git.rb:317:in `try_resolve_remote_head_from_server'
  WARN:   /home/autoproj/.local/share/autoproj/gems/ruby/2.7.0/gems/autobuild-1.23.1/lib/autobuild/import/git.rb:331:in `resolve_remote_head'
```

Can you suggest an alternative fix for that?

Reply:
... figure out why there are so many open files ? The daemon is single-threaded and does not keep any long-lived connection / open files. Sounds pretty weird to me that you're having a problem with this. Maybe it is leaking some file descriptors ?

Reply:
> Maybe it is leaking some file descriptors ?

I would think so but after a quick inspection to `Autobuild::Subcommand` I didn't find any. Either way, given the current limitations I figured avoiding hundreds of git invocations made sense regardless.

I guess this will require further investigation. I've opened another PR so we can keep track of it.